src/Server.cpp

```diff
@@ -925,6 +1106,115 @@ bool isCompleteBody(Connection& connection)
+ * If only part of the body is written, it updates the request body to contain the remaining data.
+ *
+ * @param server Reference to the Server object.
+ * @param connection Reference to the Connection object.
+ */
+void connectionSendToCGI(Server& server, Connection& connection)
+{
+	if (connection.m_request.body.empty()) {
+		LOG_ERROR << "Error: empty body: can't send to CGI";
+		connection.m_request.httpStatus = StatusInternalServerError;
+		connection.m_status = Connection::BuildResponse;
+		server.modifyEvent(connection.m_clientFd, EPOLLOUT);
+		server.removeCGIFileDescriptor(server, connection.m_pipeFromCGIReadEnd);
+		server.removeCGIFileDescriptor(server, connection.m_pipeToCGIWriteEnd);
+		return;
+	}
+
+	long bytesSent
+		= write(connection.m_pipeToCGIWriteEnd, connection.m_request.body.c_str(), connection.m_request.body.size());
+
+	if (bytesSent == -1) {
+		LOG_ERROR << "Error: write(): can't send to CGI: " + std::string(std::strerror(errno));
+		connection.m_request.httpStatus = StatusInternalServerError;
+		connection.m_status = Connection::BuildResponse;
+		server.modifyEvent(connection.m_clientFd, EPOLLOUT);
+		server.removeCGIFileDescriptor(server, connection.m_pipeFromCGIReadEnd);
+		server.removeCGIFileDescriptor(server, connection.m_pipeToCGIWriteEnd);
+		return;
+	}
+	if (bytesSent == static_cast<long>(connection.m_request.body.size())) {
+		LOG_DEBUG << "CGI: Full body sent";
+		connection.m_status = Connection::ReceiveFromCGI;
+		connection.m_request.body.clear();
+		server.removeCGIFileDescriptor(server, connection.m_pipeToCGIWriteEnd);
+		return;
+	}
+	connection.m_request.body = connection.m_request.body.substr(bytesSent);
+}
+
+/**
+ * @brief Handles receiving data from a CGI script and processes it accordingly.
+ *
+ * This function reads data from the CGI script's output pipe and appends it to the request body.
+ * It also handles errors during the read operation and manages the connection state based on the
+ * data received. If the read operation indicates that the CGI script has finished sending data,
+ * the function will transition the connection to the response-building state and clean up the
+ * associated file descriptors.
+ *
+ * @param server Reference to the Server instance managing the connection.
+ * @param connection Reference to the Connection instance representing the client connection.
+ */
+void connectionReceiveFromCGI(Server& server, Connection& connection)
+{
+	char buffer[Server::s_cgiBodyBufferSize] = {};
+	long bytesRead = read(connection.m_pipeFromCGIReadEnd, buffer, sizeof(buffer));
+
+	if (bytesRead == -1) {
+		LOG_ERROR << "Error: read(): can't read from CGI: " + std::string(std::strerror(errno));
+		connection.m_request.httpStatus = StatusInternalServerError;
+		connection.m_status = Connection::BuildResponse;
+		server.modifyEvent(connection.m_clientFd, EPOLLOUT);
+		server.removeCGIFileDescriptor(server, connection.m_pipeFromCGIReadEnd);
+		if (connection.m_pipeToCGIWriteEnd != -1)
```

Comment:
Shouldn't this always be -1?

Reply:
It wouldn't be, if the CGI call only involves getting the body of the CGI script (GET method). The pipeToCGIWriteEnd would still be open and therefore not -1.

Reply:
Now only opens pipeToCGI if the calling method is POST. Added wrapper for closing pipeEnds.